src/Sema.zig

```diff
@@ -6294,6 +6294,71 @@ fn zirBitcast(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air
     const extra = sema.code.extraData(Zir.Inst.Bin, inst_data.payload_index).data;
 
     const dest_ty = try sema.resolveType(block, dest_ty_src, extra.lhs);
+    switch (dest_ty.zigTypeTag()) {
+        .Type,
+        .Void,
+        .NoReturn,
+        .ComptimeFloat,
+        .ComptimeInt,
+        .Undefined,
+        .Null,
+        .Optional,
+        .ErrorUnion,
+        .ErrorSet,
+        .Opaque,
+        .Frame,
+        .AnyFrame,
+        .EnumLiteral,
+        .Union,
+        => return sema.fail(block, dest_ty_src, "invalid type '{}' for @bitCast", .{dest_ty}),
+
+        .Pointer => {
+            const msg = msg: {
+                const msg = try sema.errMsg(block, dest_ty_src, "cannot @bitCast to pointer type '{}'", .{dest_ty});
+                errdefer msg.destroy(sema.gpa);
+
+                const pointee_ty = dest_ty.ptrInfo().data.pointee_type;
+                try sema.errNote(block, dest_ty_src, msg, "to cast to a pointer type, use @ptrCast({}, ...)", .{dest_ty});
+                try sema.errNote(block, dest_ty_src, msg, "to cast to a non-pointer type, use @bitCast({}, ...)", .{pointee_ty});
+                break :msg msg;
+            };
+            return sema.failWithOwnedErrorMsg(msg);
+        },
+
+        .Fn,
+        .BoundFn,
+        => return sema.fail(block, dest_ty_src, "cannot @bitCast to '{}', use @ptrCast to cast to a function", .{dest_ty}),
+
+        .Struct => {
+            if (dest_ty.containerLayout() == .Auto) {
+                const msg = msg: {
+                    const msg = try sema.errMsg(
+                        block,
+                        dest_ty_src,
+                        "cannot @bitCast to '{}', struct does not have a specified layout",
+                        .{dest_ty},
+                    );
+                    errdefer msg.destroy(sema.gpa);
+
+                    const ty_decl_src = dest_ty.declSrcLoc();
+                    try sema.mod.errNoteNonLazy(
+                        ty_decl_src,
+                        msg,
+                        "consider using 'packed struct' or 'extern struct' for a specified layout.",
+                        .{},
+                    );
+                    break :msg msg;
+                };
+                return sema.failWithOwnedErrorMsg(msg);
+            }
+        },
+        else => {},
```

Comment:
For readability it would be nice to explicitly list the six other valid types here.